### PR TITLE
[feature/frontend] Add options to include Unlisted posts or hide all posts

### PR DESCRIPTION
**Pull request by tsmethurst**
# Description

> If this is a code change, please include a summary of what you've coded, and link to the issue(s) it closes/implements.
>
> If this is a documentation change, please briefly describe what you've changed and why.

This pull request adds a new account setting that determines the visibility of posts that will be included in an account's web view and RSS feeds.

Closes https://github.com/superseriousbusiness/gotosocial/issues/3249
Closes https://github.com/superseriousbusiness/gotosocial/issues/900

## Checklist

Please put an x inside each checkbox to indicate that you've read and followed it: `[ ]` -> `[x]`

If this is a documentation change, only the first checkbox must be filled (you can delete the others if you want).

- [x] I/we have read the [GoToSocial contribution guidelines](https://github.com/superseriousbusiness/gotosocial/blob/main/CONTRIBUTING.md).
- [x] I/we have discussed the proposed changes already, either in an issue on the repository, or in the Matrix chat.
- [x] I/we have not leveraged AI to create the proposed changes.
- [x] I/we have performed a self-review of added code.
- [x] I/we have written code that is legible and maintainable by others.
- [x] I/we have commented the added code, particularly in hard-to-understand areas.
- [x] I/we have made any necessary changes to documentation.
- [ ] I/we have added tests that cover new code.
- [x] I/we have run tests and they pass locally with the changes.
- [x] I/we have run `go fmt ./...` and `golangci-lint run`.

### CODE REVIEW

File: internal/processing/account/update.go [hunk 1]
```diff
@@ -330,28 +278,175 @@ func (p *Processor) Update(ctx context.Context, account *gtsmodel.Account, form
...
+		return nil, gtserror.NewErrorInternalError(err)
 	}
 
-	p.state.Workers.Client.Queue.Push(&messages.FromClientAPI{
-		APObjectType:   ap.ActorPerson,
-		APActivityType: ap.ActivityUpdate,
-		GTSModel:       account,
-		Origin:         account,
-	})
+	if sendUpdate {
```
Comment (tsmethurst) [inline internal/processing/account/update.go:311]: I can change this in a separate PR if desired. I was just getting a bit annoyed with changing an internal setting and having my instance send out a bunch of unnecessary Updates.

File: internal/processing/account/update.go [hunk 2]
```diff
@@ -330,28 +278,175 @@ func (p *Processor) Update(ctx context.Context, account *gtsmodel.Account, form
...
+		return nil, gtserror.NewErrorInternalError(err)
 	}
+
 	return acctSensitive, nil
 }
 
+// updateFields sets FieldsRaw on the given
+// account, and resets account.Fields to an
+// empty slice, ready for further processing.
+func (p *Processor) updateFields(
```
Comment (tsmethurst) [inline internal/processing/account/update.go:333]: No logic changes here (and below), just extracted to a separate function for readability.

File: docs/user_guide/settings.md [hunk 3]
```diff
@@ -76,6 +76,26 @@ Some examples:
...
+
+!!! warning
+    Be aware that changes to this setting also apply retroactively.
+    
+    That is, if you previously made a post on Unlisted visibility, while set to show only Public posts on your profile, and you change this setting to show Public and Unlisted, then the Unlisted post you previously made will be visible on your profile alongside your Public posts.
+    
+    Likewise, if you change this setting to show no posts, then all your posts will be hidden from your profile, regardless of when you created them, and what this option was set to at the time. This will apply until you change this setting again.
+
+!!! tip
+    Alongside (domain-)blocking, this is a good "emergency" setting to use if you're facing harassment from people trawling through your public posts. It won't hide your posts from people who can see them in their clients, via ActivityPub, but it will at least prevent them from being able to click through your posts in their browser with no authentication, and easily share them with others with a URL.
```
Comment (NyaaaWhatsUpDoc) [inline docs/user_guide/settings.md:97]: the way this is worded, this makes it sound like the default behaviour is that we allow unlisted posts to be shown in the web view. i'd maybe reword the beginning to say something along the lines of "if you haven't already (this is the default behaviour)"

File: internal/db/bundb/migrations/20240906144432_unauthed_visibility.go.go [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+	"github.com/superseriousbusiness/gotosocial/internal/gtsmodel"
+	"github.com/uptrace/bun"
+)
+
+func init() {
+	up := func(ctx context.Context, db *bun.DB) error {
+		return db.RunInTx(ctx, nil, func(ctx context.Context, tx bun.Tx) error {
+
+			// If column already exists we don't need to do anything.
+			exists, err := doesColumnExist(ctx, tx,
```
Comment (NyaaaWhatsUpDoc) [inline internal/db/bundb/migrations/20240906144432_unauthed_visibility.go.go:32]: this is a usefullittle function isn't it :D

File: internal/gtsmodel/accountsettings.go [hunk 5]
```diff
@@ -32,9 +35,44 @@ type AccountSettings struct {
...
 	InteractionPolicyMutualsOnly   *InteractionPolicy `bun:""`                                                            // Interaction policy to use for new mutuals only visibility statuses. If null, assume default policy.
 	InteractionPolicyFollowersOnly *InteractionPolicy `bun:""`                                                            // Interaction policy to use for new followers only visibility statuses. If null, assume default policy.
 	InteractionPolicyUnlocked      *InteractionPolicy `bun:""`                                                            // Interaction policy to use for new unlocked visibility statuses. If null, assume default policy.
 	InteractionPolicyPublic        *InteractionPolicy `bun:""`                                                            // Interaction policy to use for new public visibility statuses. If null, assume default policy.
 }
+
+type ShowWebStatuses int
+
+const (
+	ShowWebStatusesPublicOnly = iota
```
Comment (NyaaaWhatsUpDoc) [inline internal/gtsmodel/accountsettings.go:49]: i'd make this to be specified integer values here. otherwise with an iota we risk the integer values changing (and so risk changing behaviour based on existing stored values) if a new-one is inserted earlier in the list.

Comment (NyaaaWhatsUpDoc) [reply internal/gtsmodel/accountsettings.go:49]: (depends on the response to the above comment)

File: internal/gtsmodel/accountsettings.go [hunk 6]
```diff
@@ -32,9 +35,44 @@ type AccountSettings struct {
 	CustomCSS                      string             `bun:",nullzero"`                                                   // Custom CSS that should be displayed for this Account's profile and statuses.
 	EnableRSS                      *bool              `bun:",nullzero,notnull,default:false"`                             // enable RSS feed subscription for this account's public posts at [URL]/feed
 	HideCollections                *bool              `bun:",nullzero,notnull,default:false"`                             // Hide this account's followers/following collections.
+	ShowWebStatuses                ShowWebStatuses    `bun:",notnull,default:0"`                                          // Visibility level of statuses that visitors can view via the web profile.
```
Comment (NyaaaWhatsUpDoc) [inline internal/gtsmodel/accountsettings.go:38]: alternatively instead of `ShowWebStatuses`, what if we used say `WebVisibility` that uses a `StatusVisibility` as the type? means less new types, works with existing API concepts (including in the frontend for users working with our API), and means we don't need overly wordy (imo) value types like "public_and_unlisted". it also allows flexibility in the future if we add any other visibility types between public and unlisted since it basically just works as a "all statuses >= this visibility are visible in the web frontend". though obviously we probably want to set "unlisted" as the max allowable limit.

what do you think?

File: internal/processing/account/update.go (hunk 1 above)
Comment (NyaaaWhatsUpDoc) [reply internal/processing/account/update.go:311]: i think maybe a separate PR might be more useful for this if that's good. i have some other qualms with us needing to invalidate caches outside of invalidation hooks and the database itself (as the less siloed we keep that cache invalidation logic, the more likely we are to forget to invalidate a cache as it could be done *anywhere* in the codebase)

Comment (tsmethurst) [reply internal/processing/account/update.go]: Alright, I removed it for now, and moved invalidation back to the database (but only calling it if either all columns change, or that column changes): https://github.com/superseriousbusiness/gotosocial/pull/3272/commits/438224113c333895d9d1dba18e9078d7b91dec36

File: internal/db/bundb/migrations/20240906144432_unauthed_visibility.go.go (hunk 4 above)
Comment (tsmethurst) [reply internal/db/bundb/migrations/20240906144432_unauthed_visibility.go.go:32]: It really is!

File: internal/gtsmodel/accountsettings.go (hunk 6 above)
Comment (tsmethurst) [reply internal/gtsmodel/accountsettings.go]: Right, how's this? https://github.com/superseriousbusiness/gotosocial/pull/3272/commits/db141f95d78516416d266f5285317a73a05a130d

File: internal/gtsmodel/accountsettings.go (hunk 5 above)
Comment (tsmethurst) [reply internal/gtsmodel/accountsettings.go]: Changed it now to use Visibility so this isn't relevant anymore :)
